### PR TITLE
Reduce iterator key comparison for upper/lower bound check (2nd attempt)

### LINKED ISSUE #5463: Range scan optimization #5111 cause MyRocks test failure
#5111 caused MyRocks test rocksdb.optimizer_loose_index_scans to fail initially. #5428 partly revert the optimization, but MyRocks test rocksdb.group_min_max is still failing after that. The patch is reverted by #5440. Open an issue to track.

### Expected behavior
The optimization should not fail MyRocks test.

### Actual behavior
Both rocksdb.optimizer_loose_index_scans and rocksdb.group_min_max fails with the patch.

### Steps to reproduce the behavior
Run the MyRocks tests.

### PR DESCRIPTION
**Pull request by yiwu-arbug**
Summary:
This is a second attempt for #5111, with the fix to redo iterate bounds check after `SeekXXX()`. This is because MyRocks may change iterate bounds between seek.

See #5111 for original benchmark result and discussion.

Closes #5463.

Test Plan:
Existing rocksdb tests, plus myrocks test `rocksdb.optimizer_loose_index_scans` and `rocksdb.group_min_max`.

### CODE REVIEW

**Comment by yiwu-arbug** [comment]
@ltamasi @siying mind take a look? Thanks.

**Comment by yiwu-arbug** [comment]
@siying sure. By the way, is changing iterate bound before seek an officially supported API? Thanks.

**Comment by siying** [comment]
@yiwu-arbug I know LogDevice is using it. MyRocks is probably using it too.

**Comment by yiwu-arbug** [comment]
test added.

**Comment by siying** [comment]
Our test internal hang for 2 days. Let me cancel it and retry.

**Comment by jsteemann** [comment]
@yiwu-arbug @siying : I hope I am overlooking something, but right now my understanding is that this PR has been closed but not merged back into RocksDB `master`.
IMHO reducing the number of bounds checks is quite an important optimization given that there can be expensive Comparator functions. Are there any further attempts at addressing this issue (again)?
Thanks!

**Comment by yiwu-arbug** [comment]
oops, I thought it is merged.

**Comment by yiwu-arbug** [comment]
@siying Do you mind retry landing this PR? Thanks.

File: db/version_set.cc [hunk 1]
```diff
@@ -955,6 +970,7 @@ class LevelIterator final : public InternalIterator {
       smallest_compaction_key = (*compaction_boundaries_)[file_index_].smallest;
       largest_compaction_key = (*compaction_boundaries_)[file_index_].largest;
     }
+    CheckMayBeOutOfLowerBound();
```
Comment (siying) [inline db/version_set.cc]: @yiwu-arbug I'm investigating an assert failure and trying to understand the code here.
Here CheckMayBeOutOfLowerBound() is called, and inside, it checks Valid() first. Valid() can be false for whatever the previous file iterator's state is. If it is false, then may_be_out_of_lower_bound_ is not updated. Did I miss anything.

Here is attached one issue I see. The assertion failure:
```
db/db_iter.cc:866: void rocksdb::DBIter::PrevInternal(): Assertion `iterate_lower_bound_ == nullptr || iter_.MayBeOutOfLowerBound() || user_comparator_.Compare(saved_key_.GetUserKey(), *iterate_lower_bound_) >= 0' failed.
```
Here is what is odd:
```
LevelIterator.may_be_out_of_lower_bound_ = false

LevelIterator.read_options_.iterate_lower_bound points to "0x00    0x00    0x00    0x00    0x04    0xbb    0x7a    0x07"

flevel_->files[file_index_].smallest_key points to "0x00    0x00    0x00    0x00    0x03    0xae    0x7f    0x99"
```
This indicates that LevelIterator.may_be_out_of_lower_bound_ is a wrong value.

Comment (yiwu-arbug) [reply db/version_set.cc]: I think what you point out about the valid check make sense. How about I send a fix tomorrow and see if it fixes stress test?

**Pull request closed without merging**